blivet/fstab.py

```diff
@@ -0,0 +1,792 @@
+        self.dest_file = dest_file
+
+        # prefered spec identification type; default "UUID"
+        # possible values: None, "UUID", "LABEL", "PARTLABEL", "PARTUUID", "PATH"
+        self.spec_type = None
+
+        if self.src_file is not None:
+            self.read(self.src_file)
+
+    def __deepcopy__(self, memo):
+        clone = FSTabManager(src_file=self.src_file, dest_file=self.dest_file)
+        clone._table = Table()
+        clone._table.enable_comments(True)
+
+        # Two special variables for the first and last comment. When assigning None to them, libmount fails.
+        # Notice that we are copying the value from the instance of the same type.
+        if self._table.intro_comment is not None:
+            clone._table.intro_comment = self._table.intro_comment
+        if self._table.trailing_comment is not None:
+            clone._table.trailing_comment = self._table.trailing_comment
+
+        entry = self._table.next_fs()
+        entries = []
+        while entry:
+            entries.append(entry)
+            entry = self._table.next_fs()
+
+        for entry in entries:
+            new_entry = self._copy_fs_entry(entry)
+            clone._table.add_fs(new_entry)
+
+        return clone
+
+    def __repr__(self):
+        entry = self._table.next_fs()
+        entries_str = ""
+        while entry:
+            entries_str += repr(FSTabEntry(entry=entry)) + '\n'
+            entry = self._table.next_fs()
+        return entries_str
+
+    def __iter__(self):
+        return FSTabManagerIterator(self)
+
+    def _copy_fs_entry(self, entry):
+        """ Create copy of libmount.Fs()
+        """
+        # Nope, it has to be done like this. Oh well...
+        new_entry = Fs()
+        new_entry.source = entry.source
+        new_entry.target = entry.target
+        new_entry.fstype = entry.fstype
+        new_entry.append_options(entry.options)
+        new_entry.freq = entry.freq
+        new_entry.passno = entry.passno
+        if entry.comment is not None:
+            new_entry.comment = entry.comment
+        return new_entry
+
+    def _parser_errcb(self, tb, fname, line):  # pylint: disable=unused-argument
+        """ Libmount interface error reporting function
+        """
+        log.error("Fstab parse error '%s:%s'", fname, line)
+        return 1
+
+    def entry_from_device(self, device):
+        """ Generate FSTabEntry object based on given blivet Device
+
+            :keyword device: device to process
+            :type device: :class: `blivet.devices.StorageDevice`
+            :returns: fstab entry object based on device
+            :rtype: :class: `FSTabEntry`
+        """
+
+        entry = FSTabEntry()
+
+        if device.format.uuid:
+            entry.spec = "UUID=%s" % device.format.uuid
+        else:
+            entry.spec = getattr(device, "fstab_spec", None)
+
+        entry.file = None
+        if device.format.mountable:
+            entry.file = device.format.mountpoint
```

Comment:
What if the device doesn't have a mount point set? I don't think that `None` is a valid value in fstab:

```
In [13]: f.entry_from_device(nvme0n1p1)
Out[13]: UUID=3E7F-3333         efi     None    0       0
```

Reply:
`entry_from_device` is meant to produce even incomplete entries which can then be used for example to look for existing devices in fstab using `find_entry`. To prevent invalid entries to get into the output file there is a sanity check in the `FSTabManager.write`.

Reply:
This function is meant to be able to produce incomplete entries as well (these are used for looking up existing entries). There is a sanity check in `FSTabManager.write` that prevents invalid entries to be written into the file in case of misuse.

Reply:
Ok, please make a note about this in the docstring, thanks.